### DISCUSSION TITLE
request: integrate it as an addon

### DISCUSSION THREAD

**Discussion by blurymind** [Orama-Interactive/Pixelorama#562] [Ideas]
This would really make this worthwhile using over other tools, as it would allow us to create and edit sprites directly inside godot without leaving the ide

**Comment by OverloadedOrama** [comment]
I'm currently focused on adding more functionality to make it more usable, and attempt to add stuff that are important but currently missing, like Undo/Redo (see #8). After that, I will attempt to make Pixelorama a Godot plugin, because I'm definitely interested in doing that and, from what I've seen, a lot of people want that too.

**Comment by novhack** [comment]
I am going to oppose this idea. Godot IDE is already filled with a lot of docks and tabs and adding more would bring no benefits to just switching windows.
I would be more glad if Pixelorama remained as a stand alone software that anyone can use, not just Godot developers.

**Comment by OverloadedOrama** [comment]
> I am going to oppose this idea. Godot IDE is already filled with a lot of docks and tabs and adding more would bring no benefits to just switching windows.
> I would be more glad if Pixelorama remained as a stand alone software that anyone can use, not just Godot developers.

I understand your point of view. I also prefer Pixelorama being a stand-alone, so everyone can use it without needing Godot. However, I would also like to see it being available as a plug-in, for those who prefer it. Making Pixelorama a plug-in is not my current focus, but I will most likely experiment with it in the future, probably once the features and the UI are finalized.

**Comment by blurymind** [comment]
my thought was that the addon does not embed it, but rather launches the editor in a new window.

The addon just lets you create a new resource in a click, which triggers opening the sprite editor. It also lets you edit any animations sprites in a click.. it just adds a open with pixelorama  and create with pixelorama buttons where useful. Then pixelorama does its thing in a new window. When saving, godot will reload the sprites

**Comment by dphaldes** [comment]
I have been thinking about this for a while now. Godot plugin API is incomplete right now. So embedding the app isn't possible at all (until 4.0). Since tabs were added recently, opening existing files is relatively simple but opening a file in existing instance is not possible until we setup a server like [this](https://docs.godotengine.org/en/stable/classes/class_udpserver.html).

**Comment by jegor377** [comment]
> 
> 
> I have been thinking about this for a while now. Godot plugin API is incomplete right now. So embedding the app isn't possible at all (until 4.0). Since tabs were added recently, opening existing files is relatively simple but opening a file in existing instance is not possible until we setup a server like [this](https://docs.godotengine.org/en/stable/classes/class_udpserver.html).

We can use TCPServer. No need for UDP. :)

**Comment by blurymind** [comment]
I recently saw a video comparing godot to game maker and one of the points was that gamemaker is better for prototyping because it has an integrated pixel editor. :) if pixelorama could run inside godot, it would make it fantastic for prototyping
https://www.youtube.com/watch?v=3KKeFK0NHc8

**Comment by electronstudio** [comment]
What is current status? Anyone working on this?

**Comment by novhack** [comment]
As far as I know none of the contributors focuses on this right now. I think the current focus is on making Pixelorama somewhat feature complete as a standalone app that anyone can use even if they don't develop in Godot or make games.

**Comment by dphaldes** [comment]
I started working on this just a couple of days ago. After some testing, I found that plugins and the game use the same Autoload-singleton and Input maps. This might create some conflicts with the existing game. I haven't found a way to separate them or constraint them to the "plugin workspace" . Also , Pixelorama is very mature and uses a lot of custom Menus and similar UI which won't fit in Godot (in my opinion).  So having Pixelorama independent of Godot Editor is the way to go.

Instead we can have some sort of protocol  along with a helper plugin in Godot which will facilitate communications. This way other programs can leverage it also. For example: Other engines, say Unity , can have some sort of plugin to connect with Pixelorama.

**Comment by electronstudio** [comment]
Having it integrated would be very useful for education where students would only need to download one program and not two.

Would a layout like this not fit?  
![image](https://user-images.githubusercontent.com/7192452/99234238-2f0bce00-27ec-11eb-9518-1d1e60f33043.png)

**Comment by blurymind** [comment]
I would like it yes. It can work well as a tab imo. The key here is how seamless it is to edit/create sprites. Is it a button away or many clicks away?

**Comment by Technohacker** [comment]
In the meantime, I've made an [import plugin](https://github.com/Technohacker/godot_pixelorama_importer) that lets you use .pxo project files as images in Godot without manually exporting to PNG. All you'd need to do is save the project file and Godot will auto-import it :)

**Comment by novhack** [comment]
> In the meantime, I've made an [import plugin](https://github.com/Technohacker/godot_pixelorama_importer) that lets you use .pxo project files as images in Godot without manually exporting to PNG. All you'd need to do is save the project file and Godot will auto-import it :)

I like this approach much more. Godot makes it possible to set an external editor for scripts so I wonder if the same API can be used in the plugin to do such assignments for .pxo files. So they could be opened in Pixelorama directly from Godot's file system panel.
